lmdeploy/pytorch/kernels/triton_utils.py

```diff
@@ -0,0 +1,464 @@
+    allow_fp8e4nv = target[1] >= 89
+    max_num_imprecise_acc_default = {mni_acc_default}
+    options = build_cuda_options({cuda_opt_args}, )
+    sig_key = ({sig_key_str}, )
+    spec_key = ({spec_key_str}, )
+    constexpr_key = ({constexpr_key_str}, )
+    key = (get_cuda_version_key(), sig_key, constexpr_key, spec_key, options)
+
+    kernel = kernel_cache[device].get(key, None)
+    if kernel is None:
+        return jit_func[grid]({arg_key_str}, {cuda_opt_args}, **kwargs)
+
+    args = ({sig_name_str})
+    if callable(grid):
+        grid = grid(dict({grid_args}))
+    grid_size = len(grid)
+    grid_0 = grid[0]
+    grid_1 = grid[1] if grid_size > 1 else 1
+    grid_2 = grid[2] if grid_size > 2 else 1
+    tensormaps_info = kernel.metadata["tensormaps_info"]
+    if len(tensormaps_info) > 0:
+        kernel.run(grid_0, grid_1, grid_2, kernel.num_warps, kernel.num_ctas,
+                *kernel.cluster_dims,
+                kernel.shared, stream, kernel.function, launch_enter_hook,
+                launch_exit_hook, kernel,
+                *assemble_tensormap_to_arg(kernel.metadata["tensormaps_info"], args))
+    else:
+        kernel.run(grid_0, grid_1, grid_2, kernel.num_warps, kernel.num_ctas,
+                *kernel.cluster_dims,
+                kernel.shared, stream, kernel.function, launch_enter_hook,
+                launch_exit_hook, kernel,
+                {sig_name_str})
+
+    return kernel
+"""   # noqa: E501
+        scope = dict(
+            get_current_device=driver.get_current_device,
+            get_current_stream=driver.get_current_stream,
+            get_current_target=driver.get_current_target,
+            assemble_tensormap_to_arg=driver.assemble_tensormap_to_arg,
+            _specialization_key=self._specialization_key,
+            get_cuda_version_key=get_cuda_version_key,
+            CUDABackend=CUDABackend,
+            build_cuda_options=self.build_cuda_options,
+            jit_func=jit_func,
+            _key_of=JITFunction._key_of,
+            kernel_cache=jit_func.cache,
+            launch_enter_hook=CompiledKernel.launch_enter_hook,
+            launch_exit_hook=CompiledKernel.launch_exit_hook,
+        )
+        exec(src, scope)
+        return scope[f'_{fn.__name__}_launcher']
+
+    def __getitem__(self, grid):
+        """get item."""
+        return functools.partial(cast(Callable, self.run), grid=grid)
+
+
+@overload
+def wrap_jit_func(func: JITFunction):
+    ...
+
+
+@overload
+def wrap_jit_func(
+    *,
+    type_hint: TypeHintType = None,
+):
+    ...
+
+
+def wrap_jit_func(
+    func: JITFunction = None,
+    *,
+    type_hint: TypeHintType = None,
+):
+    """wrap jit func."""
+
+    def decorator(func: JITFunction):
+        triton_version = version.parse(triton.__version__)
+
+        if triton_version == version.parse('2.2.0'):
+            return JitFunction220Wrapper(func, type_hint)
+        if version.parse('2.2.0') < triton_version <= version.parse('2.3.1'):
```

Comment:
should we update the `requirements.txt` for triton version?

Reply:
triton2.3.0 does not have the same performance as 2.2 and 2.1, I don't want to recommend user to use such version